### PR TITLE
Upgrade to lucene-8.0.0-snapshot-31d7dfe6b1

### PR DESCRIPTION
**Pull request by nknize**
This PR upgrades the master branch to lucene-8.0.0-snapshot-31d7dfe6b1. There are several changes that need to be reviewed:

* removal of `MultiFields.getFields` and its impact on `TermVectorsService` is the main issue that needs reviewed /cc @romseygeek 

Other changes include:
* migration of Points encoding to selective indexing
* change to `TokenStreamComponents` as a final class

### CODE REVIEW

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/XLowerCaseTokenizerFactory.java [hunk 1]
```diff
@@ -20,26 +20,20 @@
...
-import org.apache.lucene.analysis.core.LowerCaseTokenizer;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.env.Environment;
 import org.elasticsearch.index.IndexSettings;
 import org.elasticsearch.index.analysis.AbstractTokenizerFactory;
-import org.elasticsearch.index.analysis.MultiTermAwareComponent;
 
-public class LowerCaseTokenizerFactory extends AbstractTokenizerFactory implements MultiTermAwareComponent {
+@Deprecated
+public class XLowerCaseTokenizerFactory extends AbstractTokenizerFactory {
```
Comment (jimczi) [inline modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/XLowerCaseTokenizerFactory.java]: Can we keep the old name ?

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/XLowerCaseTokenizerFactory.java [hunk 2]
```diff
@@ -20,26 +20,20 @@
...
 
     @Override
     public Tokenizer create() {
-        return new LowerCaseTokenizer();
-    }
-
-    @Override
-    public Object getMultiTermComponent() {
-        return this;
+        return new XLowerCaseTokenizer();
```
Comment (jimczi) [inline modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/XLowerCaseTokenizerFactory.java]: Can we use a LetterTokenizer followed by a LowerCaseFilter like explained in the deprecation javadocs ?
I don't think we need to keep an `XLowerCaseTokenizer`.

File: modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/CommonAnalysisFactoryTests.java [hunk 3]
```diff
@@ -48,7 +48,7 @@ public CommonAnalysisFactoryTests() {
         tokenizers.put("edgengram", EdgeNGramTokenizerFactory.class);
         tokenizers.put("classic", ClassicTokenizerFactory.class);
         tokenizers.put("letter", LetterTokenizerFactory.class);
-        tokenizers.put("lowercase", LowerCaseTokenizerFactory.class);
+        // tokenizers.put("lowercase", XLowerCaseTokenizerFactory.class);
```
Comment (jimczi) [inline modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/CommonAnalysisFactoryTests.java]: Why is it commented ?

File: test/framework/src/main/java/org/elasticsearch/indices/analysis/AnalysisFactoryTestCase.java [hunk 4]
```diff
@@ -77,7 +77,6 @@ private static String toCamelCase(String s) {
         .put("edgengram", MovedToAnalysisCommon.class)
         .put("keyword", MovedToAnalysisCommon.class)
         .put("letter", MovedToAnalysisCommon.class)
-        .put("lowercase", MovedToAnalysisCommon.class)
```
Comment (jimczi) [inline test/framework/src/main/java/org/elasticsearch/indices/analysis/AnalysisFactoryTestCase.java]: We still need this ? The tokenizer is deprecated, not yet removed ?

**Review by jimczi — COMMENTED** [review]
Thanks @nknize ! I left some comments

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/XLowerCaseTokenizerFactory.java (hunk 2 above)
Comment (romseygeek) [reply modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/XLowerCaseTokenizerFactory.java]: I'd hoped to do that, but unfortunately the contract of `create` demands that we return a `Tokenizer`, and I don't think there's an easy way of wrapping a `Tokenizer` + `TokenFilter` combination here?

File: modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/CommonAnalysisFactoryTests.java (hunk 3 above)
Comment (romseygeek) [reply modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/CommonAnalysisFactoryTests.java]: The tests here are explicitly checking that we can load lucene analysis classes.  `LowercaseTokenizer` isn't there any more, so this needs to be removed - the commenting out was just to get tests to pass.

File: test/framework/src/main/java/org/elasticsearch/indices/analysis/AnalysisFactoryTestCase.java (hunk 4 above)
Comment (romseygeek) [reply test/framework/src/main/java/org/elasticsearch/indices/analysis/AnalysisFactoryTestCase.java]: Again, this is checking for lucene classes that no longer exist.

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/XLowerCaseTokenizerFactory.java (hunk 2 above)
Comment (jimczi) [reply modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/XLowerCaseTokenizerFactory.java]: As discussed with @romseygeek offline we'll handle the deprecation and removal in a follow up pr.

**Review by jimczi — APPROVED** [review]
@nknize I pushed a fix for the failing tests and added a norelease comment regarding the LowercaseTokenizer deprecation/removal. We'll handle the deprecation/removal in a follow up to not block this pr. I am +1 to merge as is if the CI passes

**Review by s1monw — APPROVED** [review]
LGTM 2

**Comment by nknize** [comment]
Thanks @jimczi 

I'm not familiar with the error just thrown:

```
11:25:39   1> [2018-11-05T17:25:37,187][WARN ][o.e.b.JNANatives         ] [[SUITE-AnnotatedTextHighlighterTests-seed#[496176D107CD889E]]] unable to install syscall filter: 
11:25:39   2> REPRODUCE WITH: ./gradlew :plugins:mapper-annotated-text:test -Dtests.seed=496176D107CD889E -Dtests.class=org.elasticsearch.search.highlight.AnnotatedTextHighlighterTests -Dtests.method="testAnnotatedTextStructuredMatch" -Dtests.security.manager=true -Dtests.locale=de-LU -Dtests.timezone=Etc/GMT+10 -Dcompiler.java=11 -Druntime.java=8
11:25:39   1> java.lang.UnsupportedOperationException: seccomp unavailable: CONFIG_SECCOMP not compiled into kernel, CONFIG_SECCOMP and CONFIG_SECCOMP_FILTER are needed
11:25:39   1> 	at org.elasticsearch.bootstrap.SystemCallFilter.linuxImpl(SystemCallFilter.java:342) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
```

**Review by jimczi — COMMENTED** [review]
I fixed the failing tests, let's see if the CI passes now.

**Comment by nknize** [comment]
Looks like a new failure. Reproduces for me.

```14:52:46   2> REPRODUCE WITH: ./gradlew :plugins:mapper-annotated-text:integTestRunner -Dtests.seed=638D69B8A84F3208 -Dtests.class=org.elasticsearch.index.mapper.annotatedtext.AnnotatedTextClientYamlTestSuiteIT -Dtests.method="test {yaml=mapper_annotatedtext/10_basic/annotated highlighter on annotated text}" -Dtests.security.manager=true -Dtests.locale=is -Dtests.timezone=Africa/Blantyre -Dcompiler.java=11 -Druntime.java=8```

```
   > Throwable #1: java.lang.AssertionError: Failure at [mapper_annotatedtext/10_basic:38]: field [hits.hits.0.highlight.text.0] is null
   >    at __randomizedtesting.SeedInfo.seed([638D69B8A84F3208:EBD9566206B35FF0]:0)
   >    at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:407)
   >    at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.test(ESClientYamlSuiteTestCase.java:384)
   >    at java.lang.Thread.run(Thread.java:748)
   > Caused by: java.lang.AssertionError: field [hits.hits.0.highlight.text.0] is null
   >    at org.elasticsearch.test.rest.yaml.section.MatchAssertion.doAssert(MatchAssertion.java:79)
   >    at org.elasticsearch.test.rest.yaml.section.Assertion.execute(Assertion.java:76)
   >    at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:400)
   >    ... 38 more
```

**Comment by nknize** [comment]
looks like all tests passed but CentOS packaging is angry. Isn't that usually a flaky one anyway?

**Comment by jimczi** [comment]
@nknize I merged this pr and restored https://github.com/elastic/elasticsearch/pull/35225 in 6x.

**Pull request merged**